In [4]:
import pandas as pd
import statsmodels.formula.api as sms
import matplotlib.pyplot as plt
import numpy as np

boston = pd.DataFrame.from_csv("https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv")

credit = pd.read_csv("http://www-bcf.usc.edu/~gareth/ISL/Credit.csv")
credit = credit.drop(columns=['Unnamed: 0'])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  


<h1>Exercício 1</h1>

In [5]:
model_fit1 = sms.ols(formula="medv~lstat", data=boston).fit()
model_fit2 = sms.ols(formula="medv~age", data=boston).fit()
model_fit3 = sms.ols(formula="medv~age+lstat", data=boston).fit()
print(model_fit1.summary(), model_fit2.summary(), model_fit3.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.544
Model:                            OLS   Adj. R-squared:                  0.543
Method:                 Least Squares   F-statistic:                     601.6
Date:                Wed, 13 Mar 2019   Prob (F-statistic):           5.08e-88
Time:                        22:37:04   Log-Likelihood:                -1641.5
No. Observations:                 506   AIC:                             3287.
Df Residuals:                     504   BIC:                             3295.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     34.5538      0.563     61.415      0.0

modelo com lstat explica mais da feature medv (r2), embora ambos sejam altamente relacionados a medv (p-value) 
ambas features tem uma correlação negativa em relação a medv, ou seja quanto menor age e lstat, maior medv

In [6]:
model_fit4 = sms.ols(formula="medv~lstat+age+zn+indus+chas+nox+rm+dis+rad+tax+ptratio+b", data=boston).fit()
print(model_fit4.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.735
Model:                            OLS   Adj. R-squared:                  0.728
Method:                 Least Squares   F-statistic:                     113.9
Date:                Wed, 13 Mar 2019   Prob (F-statistic):          1.23e-133
Time:                        22:37:39   Log-Likelihood:                -1504.3
No. Observations:                 506   AIC:                             3035.
Df Residuals:                     493   BIC:                             3090.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     35.3797      5.143      6.879      0.0

criando um modelo com todas as variáveis é possível notar que age, diferentemente do que visto em um modelo simples, não influencia medv (p-value > 0.05), bem como indus. ambas podem ser removidas do modelo sem perda de acurácia

In [7]:
model_fit5 = sms.ols(formula="medv~lstat+zn+chas+nox+rm+dis+rad+tax+ptratio+b", data=boston).fit()
print(model_fit5.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.735
Model:                            OLS   Adj. R-squared:                  0.729
Method:                 Least Squares   F-statistic:                     137.2
Date:                Wed, 13 Mar 2019   Prob (F-statistic):          9.95e-136
Time:                        22:37:59   Log-Likelihood:                -1504.4
No. Observations:                 506   AIC:                             3031.
Df Residuals:                     495   BIC:                             3077.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     35.2255      5.107      6.898      0.0

sem age e indus r2 ajustado melhorou levemente 0.01%

<h1>Exercício 2</h1>

In [8]:
model_fit = sms.ols(formula="Balance~Gender", data=credit).fit()
print(model_fit.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1836
Date:                Wed, 13 Mar 2019   Prob (F-statistic):              0.669
Time:                        22:38:48   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6043.
Df Residuals:                     398   BIC:                             6051.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        529.5362     31.988     16.

p-value maior que 0.05 <br>
não é possível rejeitar a hipótese nula, logo não há associação entre Balance e Gender<br>
demais coeficientes não foram sequer analisados

In [9]:
model_fit = sms.ols(formula="Balance~Ethnicity", data=credit).fit()
print(model_fit.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.005
Method:                 Least Squares   F-statistic:                   0.04344
Date:                Wed, 13 Mar 2019   Prob (F-statistic):              0.957
Time:                        22:39:45   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6045.
Df Residuals:                     397   BIC:                             6057.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                531

p-value maior que 0.05 <br>
não é possível rejeitar a hipótese nula, logo não há associação entre Balance e Gender<br>
demais coeficientes não foram sequer analisados

In [10]:
model_fit = sms.ols(formula="Balance~Income+Limit+Rating+Cards+Age+Education+Gender+Student+Married+Ethnicity", data=credit).fit()
print(model_fit.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.955
Model:                            OLS   Adj. R-squared:                  0.954
Method:                 Least Squares   F-statistic:                     750.3
Date:                Wed, 13 Mar 2019   Prob (F-statistic):          1.11e-253
Time:                        22:40:25   Log-Likelihood:                -2398.7
No. Observations:                 400   AIC:                             4821.
Df Residuals:                     388   BIC:                             4869.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept               -489

p-value: maior que 0.05, para diversas features, existem relações onde podemos descartar hipótese nula<br>
As features Student, Income, Limit e Cards são as mais correlatas e que mais influenciam no modelo.<br>
R-squared ajustado: valor alto, significa que boa parte dos dadosé explicado pelo modelo. 95%<br>
Slope positivo para a maior parte das features consideradas relevantess(p-value = 0)<br>
significa que na medida que feature x cresce, Balance também cresce. Exemplo: em média, para cada USD 100 a mais de Limit, Balance cresce USD 19. Exceção: Income. para cada USD 1 a mais de Income, Balance diminui USD 8 <br>
<br>
Tiraria as features Education, Gender, Age, Married e Ethnicity, que não explicam praticamente nada dos dados

In [11]:
model_fit = sms.ols(formula="Balance~Income+Limit+Rating+Cards+Age+Student", data=credit).fit()
print(model_fit.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.955
Model:                            OLS   Adj. R-squared:                  0.954
Method:                 Least Squares   F-statistic:                     1380.
Date:                Wed, 13 Mar 2019   Prob (F-statistic):          1.58e-260
Time:                        22:41:54   Log-Likelihood:                -2400.5
No. Observations:                 400   AIC:                             4815.
Df Residuals:                     393   BIC:                             4843.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       -493.7342     24.825    -19.

Mesmo tirando todos as features cujo p-value é maior que 0.05, R-squared ajustado permanece 0.954, o mesmo com todas as variaveis. Tentativas de retirar Rating e Age (features cujo p-value sáo maiores que 0.000 pioraram o modelo) <br>
hipótese: Existe uma variabilidade nos dados que não pode ser explicada ou não faz parte das features disponibilizadas